### DISCUSSION TITLE
System Clock source in project using 'target' option

### DISCUSSION THREAD

**Discussion by the-risk-taker** [modm-io/modm#870] [Q&A] [answered]
Hi,

I have project with following configuration (I'am targeting the same MCU as on NUCLEO-F103RB board):

```xml
<library>
   <options>
      <option name=":target">stm32f103rbt6</option>
      <option name="modm:build:openocd.cfg">openocd.cfg</option>
      <option name="modm:build:cmake:include_cmakelists">False</option>
      <option name="modm:build:scons:include_sconstruct">False</option>
   </options>
   <modules>
      <module>modm:build:scons</module>
      <module>modm:build:cmake</module>
      <module>modm:platform:core</module>
      <module>modm:platform:gpio</module>
      <module>modm:platform:timer:2</module>
   </modules>
   <repositories>
      <repository><path>../submodules/modm/repo.lb</path></repository>
   </repositories>
</library>
```

I want to play around with timers and interrupts - exactly to use the `setPeriod()` method:

```cpp
Timer2::setPeriod(uint32_t microseconds, bool autoApply = true)
```

But setPeriod() method expects the `SystemClock` template parameter - from where or how can I provide the system clock value? 

**[Updated]**
I just figured it out that if I provide a struct like one from `modm/src/modm/board/nucleo_f103rb/board.hpp` for NUCELO board in my main.cpp:

```cpp
struct SystemClock
{
    static constexpr uint32_t Frequency = 72'000'000;
    static constexpr uint32_t Ahb = Frequency;
    static constexpr uint32_t Apb1 = Frequency / 2;
    static constexpr uint32_t Apb2 = Frequency;
// ...

    static bool inline enable()
    {
        // ...
    }
}
```

Then I can use  `setPeriod()` method as follows:

```cpp
Timer2::setPeriod<SystemClock>(period);
```

(1) Is this the correct approach? Or is there a better way?

(2) Maybe I don't know how to use the documentation, but the documentation didn't help me at all. Can anyone tell me where I can find this topic in the documentation?

**Comment by salkinium** [comment, marked as answer]
> (1) Is this the correct approach? Or is there a better way?

That's how it is intended to be used, you can also define multiple SystemClock structs to switch them at runtime (at the cost of calling `Peripheral::initialize<SystemClock123>()` for everything you use anytime you switch).

> Can anyone tell me where I can find this topic in the documentation?

The Timer API is simply lacking documentation, sorry about that!

**Reply by salkinium** [reply]
I intended for the SystemClock to be autogenerated using the [machine-readable clock tree data also found in CubeMX](https://salkinium.com/stm32/clock/), but [it was more complicated than I thought](https://github.com/salkinium/save-the-clocktrees) and there were just easier and more important things to do in the meantime…

**Reply by rleh** [reply]
As an intermediate step we could seperate the hand-coded `SystemClock` structs from the `:board:*` modules and put them in there own lbuild modules so that the code can be reused in Custom board support packages (instead of being copy-pasted).
Later these lbuild modules can be extended with autogenerated clocks, multiple options, ...
